tests/mechanic/launcher_test.py

```diff
@@ -119,6 +119,7 @@ def wait(self):
 
 
 class ProcessLauncherTests(TestCase):
+    @mock.patch('os.path.join', return_value="/telemetry")
```

Comment:
Why do we need to mock this? IMHO we shouldn't have mocked quite a few of the dependencies of this test case to begin with and specifically not `esrally.mechanic.provisioner.NodeConfiguration`. 

I think we need to clean this up in a separate PR. @drawlerr can you please work on a new PR (based on master) and rewrite this test case to mock only dependencies that have side effects? (as a guideline: things in `os` are likely ok, things like `esrally.config.Config` or `esrally.metrics.MetricsStore` should not be mocked).

Reply:
I modified the launcher to use os.path instead of concatenating strings. If I don't update the test I get the following error. Not sure how to get away without mocking it..
```
________________ ProcessLauncherTests.test_daemon_start_stop __________________

args = (<tests.mechanic.launcher_test.ProcessLauncherTests testMethod=test_daemon_start_stop>, <MagicMock name='Process' id='...ion' id='4573833480'>, <MagicMock name='MetricsStore' id='4573862824'>, <MagicMock name='Config' id='4577452888'>, ...)
keywargs = {}
extra_args = [<MagicMock name='Process' id='4574868424'>, <MagicMock name='wait_for_pidfile' id='4578813376'>, <MagicMock name='Nod...tricsStore' id='4573862824'>, <MagicMock name='Config' id='4577452888'>, <MagicMock name='chdir' id='4578781056'>, ...]
entered_patchers = [<unittest.mock._patch object at 0x10d987438>, <unittest.mock._patch object at 0x10d987ef0>, <unittest.mock._patch obj...bject at 0x10d987a58>, <unittest.mock._patch object at 0x10d987320>, <unittest.mock._patch object at 0x10d987ac8>, ...]
exc_info = (<class 'TypeError'>, TypeError("test_daemon_start_stop() missing 1 required positional argument: 'path'"), <traceback object at 0x110efe5c8>)
patching = <unittest.mock._patch object at 0x10d987438>
arg = <MagicMock name='kill' id='4578728480'>

    @wraps(func)
    def patched(*args, **keywargs):
        extra_args = []
        entered_patchers = []

        exc_info = tuple()
        try:
            for patching in patched.patchings:
                arg = patching.__enter__()
                entered_patchers.append(patching)
                if patching.attribute_name is not None:
                    keywargs.update(arg)
                elif patching.new is DEFAULT:
                    extra_args.append(arg)

            args += tuple(extra_args)
>           return func(*args, **keywargs)
E           TypeError: test_daemon_start_stop() missing 1 required positional argument: 'path'

/usr/local/Cellar/python/3.7.1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/mock.py:1191: TypeError
```

Reply:
You also need to remove `path` from the argument list of the test. Otherwise the test framework cannot call `test_daemon_start_stop`.

Reply:
right, sorry if if remove both mock and path in the arguments I get this:
```
    @mock.patch('os.kill')
    @mock.patch('subprocess.Popen',new=MockPopen)
    @mock.patch('esrally.mechanic.java_resolver.java_home', return_value=(12, "/java_home/"))
    @mock.patch('esrally.utils.jvm.supports_option', return_value=True)
    @mock.patch('os.chdir')
    @mock.patch('esrally.config.Config')
    @mock.patch('esrally.metrics.MetricsStore')
    @mock.patch('esrally.mechanic.provisioner.NodeConfiguration')
    @mock.patch('esrally.mechanic.launcher.wait_for_pidfile', return_value=MOCK_PID_VALUE)
    @mock.patch('psutil.Process')
    def test_daemon_start_stop(self, process, wait_for_pidfile, node_config, ms, cfg, chdir, supports, java_home, kill):
        proc_launcher = launcher.ProcessLauncher(cfg, ms, paths.races_root(cfg))

>       nodes = proc_launcher.start([node_config])

tests/mechanic/launcher_test.py:135:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
esrally/mechanic/launcher.py:294: in start
    return [self._start_node(node_configuration, node_count_on_host) for node_configuration in node_configurations]
esrally/mechanic/launcher.py:294: in <listcomp>
    return [self._start_node(node_configuration, node_count_on_host) for node_configuration in node_configurations]
esrally/mechanic/launcher.py:302: in _start_node
    node_telemetry_dir = os.path.join(node_configuration.node_root_path, "telemetry")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

a = <MagicMock name='NodeConfiguration.node_root_path' id='4598523160'>
p = ('telemetry',)

    def join(a, *p):
        """Join two or more pathname components, inserting '/' as needed.
        If any component is an absolute path, all previous path components
        will be discarded.  An empty last part will result in a path that
        ends with a separator."""
>       a = os.fspath(a)
E       TypeError: expected str, bytes or os.PathLike object, not MagicMock

/usr/local/Cellar/python/3.7.1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/posixpath.py:80: TypeError
```

Reply:
As discussed in another channel this is the result of mocking other dependencies so it is ok if you keep the mock on `os.path.join` here as this will be cleaned up in a dedicated PR.